discussions/api.py

```diff
@@ -267,46 +265,58 @@ def remove_from_channel(channel_name, discussion_username):
-
-    membership_channel_ids = set()
-    for enrollment_id in user.programenrollment_set.values_list("id", flat=True):
-        queries = search_percolate_queries(enrollment_id, PercolateQuery.DISCUSSION_CHANNEL_TYPE)
-        membership_channel_ids.update(queries.values_list("channel__id", flat=True))
-    # Make sure we don't include extra channels beyond what's in all_channel_ids
-    membership_channel_ids = membership_channel_ids.intersection(all_channel_ids)
-
-    membership_channels = Channel.objects.filter(id__in=membership_channel_ids)
-    for channel in membership_channels:
-        add_to_channel(channel.name, discussion_user.username)
+    program_ids_by_channel_id = {
+        channel_program.channel_id: channel_program.program_id
+        for channel_program in ChannelProgram.objects.all()
+    }
+    membership_ids = PercolateQueryMembership.objects.filter(
+        query__source_type=PercolateQuery.DISCUSSION_CHANNEL_TYPE,
+        needs_update=True
+    ).values_list("id", flat=True)
+
+    for membership_id in membership_ids:
+        with transaction.atomic():
+            membership = (
+                PercolateQueryMembership.objects
+                .filter(id=membership_id)
+                .prefetch_related('query__channels')
+                .select_for_update()
+                .first()
+            )
 
-    nonmembership_channels = Channel.objects.filter(
-        id__in=all_channel_ids.difference(membership_channel_ids)
-    )
-    for channel in nonmembership_channels:
-        remove_from_channel(channel.name, discussion_user.username)
+            channel = membership.query.channels.first()
+
+            # if we can't look up the program, skip this one
+            if channel is None or channel.id not in program_ids_by_channel_id:
+                continue
+
+            # if the user is a moderator, don't manipulate subscriptions
+            if Role.objects.filter(
+                    role__in=Role.permission_to_roles[Permissions.CAN_CREATE_FORUMS],
+                    user_id=membership.user_id,
+                    program_id=program_ids_by_channel_id[channel.id],
+            ).exists():
+                membership.needs_update = False
+                membership.save()
+            else:
+                try:
+                    # This guards against a race condition where the user's profile is in a celery task
+                    # and hasn't yet actually been created
+                    discussion_user = create_or_update_discussion_user(membership.user_id)
+
+                    if membership.is_member:
+                        add_to_channel(channel.name, discussion_user.username)
+                    else:
+                        remove_from_channel(channel.name, discussion_user.username)
+
+                    membership.needs_update = False
+                    membership.save()
+                except DiscussionUserSyncException:
```

Comment:
Can you add a test for this case? Just to make sure an exception doesn't propagate